Move mbed-stubs-platform to the platform directory

rwalton-arm opened the pull request:
<!--
For more information on the requirements for pull requests, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html).

NOTE: Do not remove any of the template headings (even for optional sections) as this
template is automatically parsed. 
-->

### Summary of changes <!-- Required -->

<!-- 
    Please provide the following information: 

    Description of the the change (what is this fixing / adding / removing?).

    Why the change is needed (if this is fixing a reported issue please summarize what
    the issue is and add the reference. E.g. Fixes #17119).

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
    
-->

Fixes https://github.com/ARMmbed/mbed-os/issues/14795

This PR makes the mbed-stubs-platform library more self-contained by 
removing the dependencies on the monolithic header libraries we created to 
make implementing the stubs easier. We also move the files to the platform
directory so the stubs live with the associated component.

Now the mbed-stubs-platform library should only depend on the headers it 
uses. It also now owns and exposes it's own public headers rather than 
depending on  every available stub header from the mbed-stubs-headers library.

This change makes the flow of dependencies easier to see and allows for a 
more modular architecture moving forward.

#### Impact of changes <!-- Optional -->
<!-- 
    If there are any implications for users taking this change then they must be 
    provided here. For Major PR types this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

#### Migration actions required <!-- Optional -->
<!-- 
    This should only be applicable in Major PR types for which this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

### Documentation <!-- Required -->

<!-- 
    Please provide details of any document updates required, including links to any
    related PRs against the docs repository.
    If no document updates are required please specify 'None', this at least tells us
    that this has been considered.
-->

----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

<!--
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front of them would change the meaning incorrectly. 
-->
    [x] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results <!-- Required -->

<!--
    Provide all the information required, listing all the testing performed. For new targets please attach full test results for all supported compilers.
-->
    [] No Tests required for this change (E.g docs only update)
    [x] Covered by existing mbed-os tests (Greentea or Unittest)
    [] Tests / results supplied as part of this PR
    
    
----------------------------------------------------------------------------------------------------------------
### Reviewers <!-- Optional -->

<!--
    Request additional reviewers with @username or @team
-->

----------------------------------------------------------------------------------------------------------------

rwalton-arm committed:
CMake: Move platform stubs to the mbed-os/platform directory

Move the platform stub library into the platform component directory.
This change is so we can avoid duplicating the mbed-os source tree in a
central UNITTESTS folder.

rwalton-arm committed:
CMake: unit-tests: Move mbed-headers-platform to platform directory

Move the header-only mbed-headers-platform library the unit test stubs
depend on into the platform component directory. This makes the platform
stubs more self contained and improves composition of the library.

rwalton-arm committed:
CMake: unit-tests: Make platform stub only depend on headers it uses

Previously the platform stub library depended on `mbed-headers`, which
is a collection of all available headers in mbed-os. To make it easier
to separate the library, only depend on the headers we're actually using.

rwalton-arm committed:
CMake: unit-tests: CircularBuffer: Remove trailing whitespace

rwalton-arm committed:
CMake: unit-tests: Split platform headers out of mbed-stubs-headers

We have a central collection of "stub headers", which makes reasoning
about dependencies rather difficult, as it forces every stub library to
depend on all available stub headers. The standard approach would be for
each stub library to expose its public headers, and its dependents to
explicitly specify a dependency on the stub library containing the
headers it needs. This is a more modular design than creating a
header-only monolith library. Move the platform stub headers from this
central library into the mbed-stubs-platform library to increase
modularity.

mbed-stubs-connectivity now depends on the mbed-stubs-platform because
it requires some headers which were moved to mbed-stubs-platform.

0xc0170 commented on platform/tests/UNITTESTS/doubles/CMakeLists.txt:
```diff
@@ -1,8 +1,17 @@
+        ${mbed-os_SOURCE_DIR}/platform/mbed-trace/include
+)
 
+add_library(mbed-stubs-platform)
```
stubs or doubles (the folder is named doubles but we talk or used stubs). Was there a reason to name the folder differently?

rwalton-arm replied:
I thought doubles was more "correct" as some of these are stubs, some are a bit more like fakes. I can revert the folder name back to stubs if we think that's better/less confusing.

0xc0170 replied:
doubles is fine then.

LDong-Arm reviewed (approved):
To me this looks like a good, minimal-impact approach. The stub headers and sources are unchanged but just moved into the platform directory, with CMake definitions adjusted accordingly and decoupled.

Let's see if others from @ARMmbed/mbed-os-core agree with this approach.

rajkan01 commented on platform/tests/UNITTESTS/doubles/CMakeLists.txt:
```diff
@@ -31,7 +31,9 @@ target_link_options(mbed-stubs-platform
 target_link_libraries(mbed-stubs-platform
-    PRIVATE
-        mbed-headers
+    PUBLIC
```
This PR approach is good and minimal-impact but Initially, we have a discussion to use PRIVATE instead of stubs libraries target link libraries with PUBLIC refer https://github.com/ARMmbed/mbed-os/pull/14285#discussion_r576287699 & https://github.com/ARMmbed/mbed-os/pull/14285#discussion_r576731572, please share your thoughts

rajkan01 commented on platform/tests/UNITTESTS/doubles/CMakeLists.txt:
```diff
@@ -17,14 +26,17 @@ target_sources(mbed-stubs-platform
 )
-
+target_include_directories(mbed-stubs-platform
+    PUBLIC
```
Adding all the platform stub headers into mbed-stubs-platform with PUBLIC  again, this will be a problem if library unittest (example ATCmdParser unittest) don't want to include stub platform headers but target link aginst mbed-stubs-platform brought those the platform stub headers. Initially, that was one of the reasons we added a mbed-stubs-headers library, it is better to add a new library mbed-stubs-platform-headers INTERFACE library with all platform stub headers refer our discussion https://github.com/ARMmbed/mbed-os/pull/14285#discussion_r576779176

rwalton-arm replied:
I'm not sure I'm seeing the issue with the example use case you've given. Why wouldn't you want the headers to be available if you specifically linked to the stub library? I could see the inverse of this being a potential problem: you want the "stub header" but want to provide another implementation yourself. However, if a stub has some "public headers" (every header in this `mbed-stubs-headers` lib is "public") it seems better to expose them as part of the stub library they belong to. Pulling all the headers out into a monolithic header-only library that requires non-obvious dependencies before you can use it seems like an anti-pattern. Also creating a separate "stub header" library for each stub seems unnecessary to me, unless there's a real issue we're solving by adding the extra library.

Maybe I'm missing something, so if you could show me an example of where this is actually an issue it would help me understand why we need to do this. It doesn't seem to cause a problem with the `ATCmdParser` test you mentioned.

LDong-Arm replied:
In the original discussion @rajkan01 linked, @paul-szczepanek-arm said

> If they are PUBLIC then if you depend on this library you also depend on the libraries it depends on.
>
> This is like a header relying on another header to include something rather than including it itself. I think it's better if things are self contained and just exposing thing they are meant to expose and not what they are using themselves.

My interpretation of the concern is, if stub libraries A and B both have PUBLIC headers, and B depends on A, and a test T directly depends on B, then A will be "inadvertently" be exposed to T, is this the concern? But in order for compilation to succeed (because of how preprocessors work), this exposure is necessary IMO - otherwise how can T `#include` something from B which then `#include` something from A? Both A and B need to be in the include path.

Another thing is, the stub headers are for unit tests only which work within mbed-os, not intended for general use by other repositories, so flexibility isn't a concern and we should prioritise simplicity in my opinion. We are not messing with any "actual" Mbed OS libraries.

(Sorry if I completely misunderstood the original concern!)

LDong-Arm commented on platform/tests/UNITTESTS/doubles/CMakeLists.txt:
```diff
@@ -17,14 +26,17 @@ target_sources(mbed-stubs-platform
 target_link_options(mbed-stubs-platform
     PRIVATE
         --coverage
 )
```
Given that `--coverage` is globally set in UNITTESTS/CMakeLists.txt, can we remove it from here?

rwalton-arm replied:
I'd assume we can as we set it on `CMAKE_CXX_FLAGS` directly in UNITTESTS/CMakeLists.txt. I'll look into removing it as part of this PR.

LDong-Arm replied:
This extra flag exists only in platform and hal (apart from UNITTESTS/CMakeLists.txt), so I assume they are leftovers or something...

LDong-Arm reviewed (dismissed):
See inline comment regarding `--coverage`

rwalton-arm committed:
CMake: unit-tests: Remove --coverage option from mbed-stubs-platform

We set the --coverage flag globally in UNITTESTS/CMakeLists.txt. We
shouldn't need to also set it on mbed-stubs-platform, so remove it.

paul-szczepanek-arm reviewed (dismissed):
I'm fine with however you want to organise stubs.

My original opposition was based on the concern that headers should not be forced onto the user in case he wants to use genuine or mock ones instead. But in the unlikely event that stubs expose headers that are conflicting with my headers then I'll just create my own stubs library.

LDong-Arm commented on UNITTESTS/stubs/CMakeLists.txt:
```diff
@@ -115,7 +105,6 @@ target_include_directories(mbed-stubs-headers
-add_subdirectory(platform)
 add_subdirectory(storage)
 
 add_library(mbed-stubs INTERFACE)
```
(Referring to the lines below)
Shall we remove mbed-stubs-platform from mbed-stubs, and let tests explicitly link mbed-stubs-platform? Similar to what's been done to the headers - only link what is used.
Once all component stubs have been refactors, it'd be nice to have no global mbed-stubs but only component stubs eventually.

rwalton-arm replied:
I recall some build failures when I tried that initially. I think we can only remove this when everything else depending on platform has been refactored. I'll take another look and see how tricky it is to untangle at this stage, now that we've merged some other stubs PRs.

LDong-Arm replied:
I think if we take it out from mbed-stubs, lots of tests need to have mbed-stubs-platform explicitly added - this would be tedious. I imagine platform is the trickiest given everything depends on it, and it might be true for hal too.
Personally I'm okay with leaving it here for now.

LDong-Arm reviewed (changes requested):
@rwalton-arm I just added a late suggestion - only thought of it after I've gone through the same refactoring process in #14862!
